### PR TITLE
[TS1.4] - Union typing

### PR DESCRIPTION
**Pull request by bluong**
Now that we have Typescript 1.4, we can make use of union typing so that we can avoid the use of anys wherever possible.

### CODE REVIEW

File: src/components/abstractComponent.ts [hunk 1]
```diff
@@ -192,15 +192,13 @@ export module Component {
...
      */
-    public renderTo(selector: String): AbstractComponent;
-    public renderTo(element: D3.Selection): AbstractComponent;
-    public renderTo(element: any): AbstractComponent {
+    public renderTo(element: String | D3.Selection): AbstractComponent {
       if (element != null) {
         var selection: D3.Selection;
-        if (typeof(element.node) === "function") {
-          selection = (<D3.Selection> element);
+        if (typeof(element) === "string") {
```
Comment (jtlan) [inline src/components/abstractComponent.ts]: Nice refactor, although this isn't quite in spec for the change.

File: src/components/plots/abstractPlot.ts [hunk 2]
```diff
@@ -85,10 +85,8 @@ export module Plot {
      * @param {any[]|Dataset} dataset dataset to add.
      * @returns {Plot} The calling Plot.
      */
-    public addDataset(key: string, dataset: Dataset): AbstractPlot;
-    public addDataset(key: string, dataset: any[]): AbstractPlot;
-    public addDataset(dataset: Dataset): AbstractPlot;
-    public addDataset(dataset: any[]): AbstractPlot;
+    public addDataset(dataset: Dataset | any[]): AbstractPlot;
+    public addDataset(key: string, dataset: Dataset | any[]): AbstractPlot;
```
Comment (jtlan) [inline src/components/plots/abstractPlot.ts]: Change the ordering on the union type here to match the JSDoc, or vice versa.

**Comment by jtlan** [comment]
This is pretty neat. Can we make sure the order in the union type and the order in the JSDoc lines up for all these?

File: src/components/abstractComponent.ts (hunk 1 above)
Comment (bluong) [inline src/components/abstractComponent.ts]: I couldve casted element and had a bit of a mess, but I felt this was easier

File: src/components/plots/barPlot.ts [hunk 3]
```diff
@@ -160,11 +160,7 @@ export module Plot {
      * @param {any} yValOrExtent The pixel y position, or range of y values.
```
Comment (jtlan) [inline src/components/plots/barPlot.ts]: Update this JSDoc?

File: src/utils/utils.ts [hunk 4]
```diff
@@ -139,12 +139,10 @@ export module _Util {
      * @param {number} count The length of the array to generate
      * @return {any[]}
      */
-    export function createFilledArray<T>(value: T, count: number): T[];
-    export function createFilledArray<T>(func: (index?: number) => T, count: number): T[];
-    export function createFilledArray<T>(value: any, count: number) {
+    export function createFilledArray<T>(value: T | ((index?: number) => T), count: number) {
```
Comment (jtlan) [inline src/utils/utils.ts]: please update above TSDoc to match (I can't comment on the actual TSDoc line for some reason).

**Comment by jtlan** [comment]
Docs.

**Pull request merged**